Source/santa-driver/SantaDecisionManager.cc

```diff
@@ -14,9 +14,51 @@
 
 #include "SantaDecisionManager.h"
 
+// This is a made-up KAUTH_FILEOP constant which represents a
+// KAUTH_VNODE_WRITE_DATA event that gets passed to SantaDecisionManager's
+// FileOpCallback method.  The KAUTH_FILEOP_* constants are defined in
+// sys/kauth.h and run from 1--7.  KAUTH_VNODE_WRITE_DATA is already defined as
+// 4 so it overlaps with the other KAUTH_FILEOP_* constants and can't be used.
+// We define KAUTH_FILEOP_WRITE as something much greater than 7.
+#define KAUTH_FILEOP_WRITE 100
+
 #define super OSObject
 OSDefineMetaClassAndStructors(SantaDecisionManager, OSObject);
 
+# pragma mark Monitoring PIDs
+
+// This keeps track of all pids associated with compiler processes.  It is defined as a global
+// variable so that the pid monitor threads can access it without needing to reference
+// our instance of SantaDecisionManager.
+static SantaCache<bool> *compiler_pid_set_ = new SantaCache<bool>(500, 5);
+
+// Function to monitor for process termination and then remove the process pid
+// from cache of compiler pids.
+static void pid_monitor(void *param, __unused wait_result_t wait_result) {
+  pid_t pid = (pid_t)(uintptr_t)param;
+  struct timespec ts = { .tv_sec = 1, .tv_nsec = 0 }; // wait 1 sec between each poll
```

Comment:
1 sec seems like a long time: a PID could be recycled and another process will then start writing files creating unwanted transitive rules. At the same time `proc_find` is an expensive call for the whole system since it acquires the global proc list lock. So we don't want to be hammering it too often.

Maybe we could use PID + HASH(proc_name or random seed) for the compiler_pid_set_ key. We could then leave the polling at 1 sec, giving us greater confidence that a PID recycle would not result in compiler status being granted to an unworthy process.

We could store the PID in the lower 32 bits and the hash in the upper 32. This way we can extract the PID easily for lookup with proc_find.

Reply:
My thinking was that it would probably take longer than 1 sec for the PIDs to wrap back around.  I think there are 100,000 of them?  Though I suppose if you had a lot of processes already running, then the number of open PIDs might be more like 90,000.  The problem with adding a hash to the key is that we won't be able to recreate the key when we need to look up the PIDs, because we don't always have access to the proc_name.  I guess we could shorten the time?  I haven't run any tests with different times to see how it might affect performance.